### DISCUSSION TITLE
query performance with JIT enabled on Postgres 12/13

### DISCUSSION THREAD

**Discussion by tmehlinger** [sqlalchemy/sqlalchemy#6829] [Usage Questions] [answered]
First, thanks for maintaining SQLAlchemy and participating in these discussions. I've been using SQLAlchemy for a very long time and I appreciate all your work immensely.

Second, I'm aware that query performance _can be_ poor with JIT enabled--I have no questions about _what_ is happening in my code. This issue captures pretty well what's going on: https://github.com/MagicStack/asyncpg/issues/530.

That said, I have an unremarkable AIOHTTP application that integrates SQLAlchemy to store data in Postgres. I'm using the asyncpg transport and the asynchronous ORM features new in 1.4. My application code looks more or less like this:

```python
# __main__.py

from .database import db

def main():
    db_uri = 'postgresql+asyncpg://...'
    db.bind(db_uri)
    
    api = Api()
    api.start()  # calls aiotthp.web.run_app 

if __name__ == '__main__':
    main()



# database.py

Model = declarative_base()

class Database:
    def __init__(self):
        self._engine = None
        self._session_factory = None
        self._session = None

    def bind(self, db_uri: URL):
        if self._engine:
            return

        self._engine = create_async_engine(db_uri, connect_args={'server_settings': {'jit': 'off'}})
        self._session_factory = sessionmaker(self._engine, class_=AsyncSession, expire_on_commit=False)
        self._session = async_scoped_session(self._session_factory, scopefunc=asyncio.current_task)

    @property
    def engine(self):
        if not self._engine:
            raise RuntimeError('engine is not bound')
        return self._engine

    @property
    def session(self):
        if not self._session:
            raise RuntimeError('not connected, cannot begin session')
        return self._session

db = Database()



# model.py

from . import Model, db

class SomeModel(Model)
    @classmethod
    def get_by_something(cls, something):
        async with db.session() as session:
            stmt = sqlalchemy.select(cls).filter_by(something=something)
            await session.execute(stmt)
            ...
```

In this application, with JIT enabled, some requests that ultimately issue queries against some of the models will perform poorly. I don't have a complete explanation for _why_ (apart from what I indicated above), but it seems to have something to do with the graph of relationships among the models in combination with enum columns defined on some models. I use eager loading in a few places, which I suspect causes havoc with the recursive query here: https://github.com/MagicStack/asyncpg/blob/master/asyncpg/introspection.py#L86.

What I don't understand is why I'm paying the performance penalty on every query. From a comment in the issue I linked above:

> Postgres does not send enough information to describe query output at the protocol level. In order to decode complex data types, like composite types, ranges, arrays, etc, asyncpg needs to introspect those types (recursively) to determine what they are. Normally, the introspection is done once per connection, so if you use a pool you only pay the cost once.

This leads me to believe asyncpg connections are _not_ being reused, hence why the performance is so poor--for every query I'm making, asyncpg issues the introspection query before any of my own. Even with JIT disabled, I would rather not be issuing the preflight introspection query for every single one of my own queries.

Clearly, I've gone astray somewhere. I've concluded that I either don't understand how the engine works (I believed it performed connection pooling), or there's a flaw in my code... or both. Any insight would be helpful.

Thank you!

**Comment by CaselIT** [comment, marked as answer]
Hi,

The engine does use pooling by default in postgres, so connections are kept open by default. The default number of connection is 5 with an overflow of 10, that may be small in async settings.
In any case I would suggest enabling logging in the pool and or engine by setting [`echo_pool`](https://docs.sqlalchemy.org/en/14/core/engines.html?highlight=create_engine#sqlalchemy.create_engine.params.echo_pool) and [`echo`](https://docs.sqlalchemy.org/en/14/core/engines.html?highlight=create_engine#sqlalchemy.create_engine.params.echo) in the create engine call

**Reply by tmehlinger** [reply]
Thank you for the suggestions. The problem turned out to be something completely different but your answer pointed me the right direction.

I was misusing sessions in a couple places in my code, using them directly and never calling `refresh()` or `close()` to return the connection to the pool. After enough requests to the application, all the connections in the pool were being held by unclosed sessions. With that resolved (by using context managers instead), the introspection query performance became a non-issue because eventually, all the connections in the pool had the introspection result cached.

Thanks again!
